micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -15,59 +15,118 @@
         this.registry = registry;
+        this.mongoCommandUtil = new MongoCommandUtil();
+        this.collectionNamesCache = new ConcurrentHashMap<>();
+        if (collectionNamesCacheMaxSize < 1) {
+            throw new IllegalArgumentException("collectionNamesCacheMaxSize must be set to a positive value");
+        }
+        this.collectionNamesCacheMaxSize = collectionNamesCacheMaxSize;
+        this.collectionNamesCacheOverflowErrorTracker = collectionNamesCacheOverflowErrorTracker;
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, String status, long elapsedTimeInNanoseconds) {
-        timerBuilder
+    private void timeCommand(
+            CommandEvent event,
+            String collection,
+            String status,
+            long elapsedTimeInNanoseconds
+    ) {
+        Timer.builder(TIMER_NAME)
+                .description(TIMER_DESC)
                 .tag("command", event.getCommandName())
                 .tag("cluster.id", event.getConnectionDescription().getConnectionId().getServerId().getClusterId().getValue())
                 .tag("server.address", event.getConnectionDescription().getServerAddress().toString())
+                .tag("collection", collection)
                 .tag("status", status)
                 .register(registry)
                 .record(elapsedTimeInNanoseconds, TimeUnit.NANOSECONDS);
     }
 
+    private void addCollectionNameToCache(CommandEvent event, String collectionName) {
+        if (collectionNamesCache.size() < collectionNamesCacheMaxSize) {
+            collectionNamesCache.put(event.getRequestId(), collectionName);
+            return;
+        }
+        if (collectionNamesCacheOverflowErrorTracker == null) {
+            return;
+        }
+        // Cache over capacity - log every ~Nth for a good balance of warning and not spamming the logs.
+        collectionNamesCacheOverflowErrorTracker.maybeInvoke(() -> LOG.warn("Collection names cache is full - Mongo is not calling listeners properly"));
```

Comment:
In practice, we have found that logging this on the very 1st occurrence only, is easy for it to "get lost" in the sea of logs. Every 100th (or Nth) occurrence is a middle-ground between spam and seeing that there is something weird going on in the system. Ideally, this should never happen to begin with.

Reply:
Oh yeh, if there is already a mechanism for this I will gladly replace.